### DISCUSSION TITLE
Help with the "precomputable_affine" Model from Thinc.api

### DISCUSSION THREAD

**Discussion by dhruvbhimani-2512** [explosion/spaCy#10492] [Help: Model Advice]
I was going through the internal code of spaCy NER model and I came across the file: [https://github.com/explosion/spaCy/blob/2eef47dd26a5acbc3f667a2bc3b1ddf16f2d1b07/spacy/ml/_precomputable_affine.py](url)

It is using "precomputable_affine" model which I couldn't find anywhere in Thinc's website. Can anybody help me understand this or share some references?

Thanks.

**Comment by thomashacker** [comment]
Hello,
the `precomputable_affine` model is defined in this [file](https://github.com/explosion/spaCy/blob/master/spacy/ml/_precomputable_affine.py) and contains the forward- , backpropagation- and initialization function. There are no further docs on this particular model.

**Reply by dhruvbhimani-2512** [reply]
Hi @thomashacker , the thing is that they are directly defining `precomputable_affine` model like 
```
model = Model(
        "precomputable_affine", ...
```

But I want to know the internal layer structure of that. How can I find them?

**Reply by thomashacker** [reply]
When initializing a thinc `Model` object, the first parameter is the identifier for this particular model/layer (https://thinc.ai/docs/api-model#init). You can find the internal structure of every model/layer in their [forward function](https://github.com/explosion/spaCy/blob/master/spacy/ml/_precomputable_affine.py#L19)

**Reply by dhruvbhimani-2512** [reply]
@thomashacker there is no documentation for `precomputable_affine` type on [page](https://thinc.ai/docs/api-model#init).

**Reply by thomashacker** [reply]
That's correct. There's no documentation in Thinc about the `precomputable_affine` layer, because it's defined in this private [spaCy module](https://github.com/explosion/spaCy/blob/master/spacy/ml/_precomputable_affine.py#L6) and is considered an implementation detail. You can use the layer on it's own but I doubt it's something you'd want.

**Reply by jaskaran-td** [reply]
@dhruvbhimani-2512 did u got any idea for precomputable_affine architecture.

**Reply by dhruvbhimani-2512** [reply]
@jaskaran-td not yet. Please do post here if you have any ref.

**Reply by jaskaran-td** [reply]
sure man.

**Reply by jaskaran-td** [reply]
@dhruvbhimani-2512 you can read this [blog](https://explosion.ai/blog/parsing-english-in-python) for the better understanding of ner.

**Reply by dhruvbhimani-2512** [reply]
Hi @jaskaran-td , this blog is about text parsing, right? Can you explain how is this related to NER?
